File: src/repotask/kb/source.py
"""Resolve the knowledge base to a directory on disk.

Two sources, one schema:
  * remote git repository  -> cloned into ~/.repo-task/kb/<slug> and pinned to a ref
  * local project directory -> used in place, no syncing

Remote wins when both are configured.
"""

from __future__ import annotations

import hashlib
import os
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from repotask.config.models import KnowledgeConfig, RepoTaskConfig
from repotask.errors import RepoTaskError
from repotask.git import run_git

CACHE_ENV = "REPOTASK_CACHE_DIR"


def cache_root() -> Path:
    override = os.environ.get(CACHE_ENV)
    return Path(override).expanduser() if override else Path.home() / ".repo-task/kb"


def slug_for(remote: str) -> str:
    """Readable directory name plus a hash, so similar remotes never collide."""
    tail = re.sub(r"[^a-zA-Z0-9]+", "-", remote.rstrip("/").split("/")[-1]).strip("-").lower()
    digest = hashlib.sha256(remote.encode("utf-8")).hexdigest()[:10]
    return f"{tail or 'kb'}-{digest}"


@dataclass(frozen=True)
class KnowledgeSource:
    path: Path
    kind: str  # "remote" | "local"
    remote: str = ""
    ref: str = ""
    revision: str = ""
    synced_at: str = ""

    @property
    def exists(self) -> bool:
        return self.path.is_dir()


def resolve(config: RepoTaskConfig, sync: bool | None = None) -> KnowledgeSource:
    """Return the knowledge base location, syncing the remote when allowed."""
    knowledge = config.knowledge
    if knowledge.remote:
        should_sync = knowledge.auto_sync if sync is None else sync
        return _resolve_remote(knowledge, should_sync)
    local = (config.root / knowledge.local).resolve()
    if not local.is_dir():
        raise RepoTaskError(
            f"Knowledge base not found at {local}. Configure `knowledge.remote` or create the "
            "directory, then run `repo-task kb sync`."
        )
    return KnowledgeSource(path=local, kind="local")


def _resolve_remote(knowledge: KnowledgeConfig, should_sync: bool) -> KnowledgeSource:
    path = cache_root() / slug_for(knowledge.remote)
    if not path.is_dir():
        if not should_sync:
            raise RepoTaskError(
                f"Knowledge base has never been synced to {path}. Run `repo-task kb sync`."
            )
        _clone(knowledge.remote, path)
    elif should_sync:
        _fetch(path)
    _checkout(path, knowledge.ref)
    return KnowledgeSource(
        path=path,
        kind="remote",
        remote=knowledge.remote,
        ref=knowledge.ref,
        revision=run_git(["rev-parse", "HEAD"], path).strip(),
        synced_at=_read_stamp(path, should_sync),
    )


def _clone(remote: str, path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    try:
        run_git(["clone", "--quiet", "--", remote, str(path)])
    except RepoTaskError as error:
        raise RepoTaskError(f"Could not clone knowledge base {remote}: {error}") from error


def _fetch(path: Path) -> None:
    try:
        run_git(["fetch", "--quiet", "--prune", "origin"], path)
    except RepoTaskError as error:
        raise RepoTaskError(f"Could not fetch knowledge base updates: {error}") from error


def _checkout(path: Path, ref: str) -> None:
    """Pin the worktree to `ref`, preferring the remote-tracking branch when one exists."""
    if ref.startswith("-"):
        raise RepoTaskError(f"Invalid knowledge base ref: {ref}")
    target = ref
    try:
        run_git(["rev-parse", "--verify", f"origin/{ref}"], path)
        target = f"origin/{ref}"
    except RepoTaskError:
        pass
    try:
        run_git(["checkout", "--quiet", "--detach", target], path)
    except RepoTaskError as error:
        raise RepoTaskError(f"Knowledge base ref not found: {ref} ({error})") from error


def _read_stamp(path: Path, synced_now: bool) -> str:
    stamp = path.parent / f"{path.name}.synced"
    if synced_now:
        value = datetime.now(timezone.utc).isoformat(timespec="seconds")
        stamp.write_text(value, encoding="utf-8")
        return value
    return stamp.read_text(encoding="utf-8").strip() if stamp.is_file() else ""

File: rust/src/kb/source.rs
//! Resolve the knowledge base to a directory on disk.
//!
//! Remote git repositories are cloned into `~/.repo-task/kb/<slug>` and pinned to a
//! ref; a local directory is used in place. Remote wins when both are configured.

use std::path::{Path, PathBuf};

use anyhow::{bail, Context, Result};
use serde::Serialize;

use crate::config::{KnowledgeConfig, RepoTaskConfig};
use crate::git;

pub const CACHE_ENV: &str = "REPOTASK_CACHE_DIR";

pub fn cache_root() -> PathBuf {
    if let Ok(override_path) = std::env::var(CACHE_ENV) {
        return PathBuf::from(override_path);
    }
    home_dir().join(".repo-task/kb")
}

fn home_dir() -> PathBuf {
    std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .map(PathBuf::from)
        .unwrap_or_else(|_| PathBuf::from("."))
}

/// Readable directory name plus a hash, so similar remotes never collide.
pub fn slug_for(remote: &str) -> String {
    let tail: String = remote
        .trim_end_matches('/')
        .rsplit('/')
        .next()
        .unwrap_or("kb")
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() {
                c.to_ascii_lowercase()
            } else {
                '-'
            }
        })
        .collect();
    let tail = tail.trim_matches('-').to_string();
    let tail = if tail.is_empty() {
        "kb".to_string()
    } else {
        tail
    };
    format!("{tail}-{}", short_hash(remote))
}

/// FNV-1a, truncated. Only needs to avoid collisions between cache directories.
fn short_hash(value: &str) -> String {
    let mut hash: u64 = 0xcbf2_9ce4_8422_2325;
    for byte in value.as_bytes() {
        hash ^= *byte as u64;
        hash = hash.wrapping_mul(0x1000_0000_01b3);
    }
    format!("{hash:016x}")[..10].to_string()
}

#[derive(Debug, Clone, Serialize)]
pub struct KnowledgeSource {
    pub path: PathBuf,
    pub kind: String,
    pub remote: String,
    pub r#ref: String,
    pub revision: String,
    pub synced_at: String,
}

pub fn resolve(config: &RepoTaskConfig, sync: Option<bool>) -> Result<KnowledgeSource> {
    let knowledge = &config.knowledge;
    if !knowledge.remote.is_empty() {
        let should_sync = sync.unwrap_or(knowledge.auto_sync);
        return resolve_remote(knowledge, should_sync);
    }
    let local = config.root.join(&knowledge.local);
    if !local.is_dir() {
        bail!(
            "Knowledge base not found at {}. Configure `knowledge.remote` or create the \
             directory, then run `repo-task kb sync`.",
            local.display()
        );
    }
    Ok(KnowledgeSource {
        path: local,
        kind: "local".into(),
        remote: String::new(),
        r#ref: String::new(),
        revision: String::new(),
        synced_at: String::new(),
    })
}

fn resolve_remote(knowledge: &KnowledgeConfig, should_sync: bool) -> Result<KnowledgeSource> {
    let path = cache_root().join(slug_for(&knowledge.remote));
    if !path.is_dir() {
        if !should_sync {
            bail!(
                "Knowledge base has never been synced to {}. Run `repo-task kb sync`.",
                path.display()
            );
        }
        clone(&knowledge.remote, &path)?;
    } else if should_sync {
        git::run(&["fetch", "--quiet", "--prune", "origin"], Some(&path))
            .context("Could not fetch knowledge base updates")?;
    }
    checkout(&path, &knowledge.r#ref)?;
    let revision = git::run(&["rev-parse", "HEAD"], Some(&path))?
        .trim()
        .to_string();
    Ok(KnowledgeSource {
        path: path.clone(),
        kind: "remote".into(),
        remote: knowledge.remote.clone(),
        r#ref: knowledge.r#ref.clone(),
        revision,
        synced_at: stamp(&path, should_sync),
    })
}

fn clone(remote: &str, path: &Path) -> Result<()> {
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    git::run(
        &["clone", "--quiet", "--", remote, &path.to_string_lossy()],
        None,
    )
    .with_context(|| format!("Could not clone knowledge base {remote}"))?;
    Ok(())
}

/// Pin the worktree to `ref`, preferring the remote-tracking branch when one exists.
fn checkout(path: &Path, reference: &str) -> Result<()> {
    if reference.starts_with('-') {
        bail!("Invalid knowledge base ref: {reference}");
    }
    let remote_ref = format!("origin/{reference}");
    let target = if git::run_ok(&["rev-parse", "--verify", &remote_ref], Some(path)) {
        remote_ref
    } else {
        reference.to_string()
    };
    git::run(&["checkout", "--quiet", "--detach", &target], Some(path))
        .with_context(|| format!("Knowledge base ref not found: {reference}"))?;
    Ok(())
}

fn stamp(path: &Path, synced_now: bool) -> String {
    let Some(parent) = path.parent() else {
        return String::new();
    };
    let name = path.file_name().unwrap_or_default().to_string_lossy();
    let marker = parent.join(format!("{name}.synced"));
    if synced_now {
        let value = now_iso();
        let _ = std::fs::write(&marker, &value);
        return value;
    }
    std::fs::read_to_string(&marker)
        .unwrap_or_default()
        .trim()
        .to_string()
}

/// UTC timestamp, second precision, without pulling in a date-time crate.
pub fn now_iso() -> String {
    let seconds = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|duration| duration.as_secs())
        .unwrap_or(0);
    let (year, month, day, hour, minute, second) = civil_from_unix(seconds as i64);
    format!("{year:04}-{month:02}-{day:02}T{hour:02}:{minute:02}:{second:02}+00:00")
}

/// Howard Hinnant's days-from-civil algorithm, inverted.
fn civil_from_unix(seconds: i64) -> (i64, u32, u32, u32, u32, u32) {
    let days = seconds.div_euclid(86_400);
    let time = seconds.rem_euclid(86_400);
    let z = days + 719_468;
    let era = z.div_euclid(146_097);
    let doe = z.rem_euclid(146_097);
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let year = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let day = (doy - (153 * mp + 2) / 5 + 1) as u32;
    let month = if mp < 10 { mp + 3 } else { mp - 9 } as u32;
    let year = if month <= 2 { year + 1 } else { year };
    (
        year,
        month,
        day,
        (time / 3600) as u32,
        ((time % 3600) / 60) as u32,
        (time % 60) as u32,
    )
}
